hoomd/mpcd/RejectionVirtualParticleFillerGPU.cuh

```diff
@@ -0,0 +1,265 @@
+                                                                 d_tmp_pos[idx].z));
+    }
+
+/*!
+ * \b implementation
+ * Using one thread per particle, we assign the particle position, velocity and tags using the compacted indices
+ * array as an input.
+ */
+__global__ void parallel_copy(unsigned int *d_compact_indices,
+                              Scalar4 *d_permanent_positions,
+                              Scalar4 *d_permanent_velocities,
+                              unsigned int *d_permanent_tags,
+                              Scalar4 *d_temporary_positions,
+                              Scalar4 *d_temporary_velocities,
+                              const unsigned int first_idx,
+                              const unsigned int first_tag,
+                              const unsigned int n_virtual,
+                              const unsigned int block_size)
+    {
+    // one thread per virtual particle
+    const unsigned int idx = blockIdx.x * blockDim.x + threadIdx.x;
+    if (idx >= n_virtual)
+        return;
+
+    // d_compact_indices holds accepted particle indices from the temporary arrays
+    unsigned int particleIdx = d_compact_indices[idx];
+    const unsigned int real_idx = first_idx + particleIdx;
+    d_permanent_positions[real_idx] = d_temporary_positions[particleIdx];
+    d_permanent_velocities[real_idx] = d_temporary_velocities[particleIdx];
+    d_permanent_tags[real_idx] = first_tag + idx;
+    }
+
+} // end namespace kernel
+
+/*!
+ * \param args Common arguments for all geometries
+ * \param geom Confined geometry
+ *
+ * \tparam Geometry type of the confined geometry \a geom
+ *
+ * \sa mpcd::gpu::kernel::draw_virtual_particles
+ */
+template<class Geometry>
+cudaError_t draw_virtual_particles(const draw_virtual_particles_args_t& args, const Geometry& geom)
+    {
+    static unsigned int max_block_size = UINT_MAX;
+    if (max_block_size == UINT_MAX)
+        {
+        cudaFuncAttributes attr;
+        cudaFuncGetAttributes(&attr, (const void*)mpcd::gpu::kernel::draw_virtual_particles<Geometry>);
+        max_block_size = attr.maxThreadsPerBlock;
+        }
+
+    unsigned int run_block_size = min(args.block_size, max_block_size);
+    dim3 grid(args.N_virt_max / run_block_size + 1);
+    mpcd::gpu::kernel::draw_virtual_particles<Geometry><<<grid, run_block_size>>>(args.d_tmp_pos, args.d_tmp_vel,
+    args.d_track_bounded_particles, args.lo, args.hi, args.first_tag, args.vel_factor, args.filler_id, args.type,
+    args.N_virt_max, args.timestep, args.seed, args.block_size, geom);
+
+    return cudaSuccess;
+    }
+
+
+cudaError_t compact_indices(const bool *d_flags,
+                            const unsigned int num_items,
+                            unsigned int *d_out,
+                            unsigned int *d_num_selected_out,
+                            unsigned int *temp_storage)
+    {
+    cub::CountingInputIterator<int> itr(0);
+    size_t temp_storage_bytes = sizeof(temp_storage);
```

Comment:
This is definitely incorrect, did this code actually run for you? `sizeof(temp_storage)` will give you the number of bytes required to store the pointer... 8 on a 64-bit operating system. Instead, you need to have a `size_t& tmp_bytes` argument (in addition to `temp_storage`). The first time this function gets called, CUB will fill in that reference with the number of bytes it needs. Then, it is your job to alloc that memory (I'll check the cc file soon), and call this function a second time.